### ISSUE TITLE
Accessibility 

### ISSUE THREAD

**Issue by tysongach** [thoughtbot/administrate#794]
I’d love for Administrate to be as accessible, as possible. We’ve done some work in that area already: https://github.com/thoughtbot/administrate/pull/51, https://github.com/thoughtbot/administrate/pull/380, https://github.com/thoughtbot/administrate/pull/410, https://github.com/thoughtbot/administrate/pull/416.

But there is more work to be done. I’m going to use this issue to capture those:

- [ ] Links must have discernible text (21 violations)
  - Target: `body > .app-container > .main-content > .search > .search__clear > a`
  - Target: `body > .app-container > .main-content > .collection-data > tbody > tr:nth-of-type(1) > .cell-data.cell-data--email > .action-show.table__link-plain` (repeat for each row in the data table)
- [ ] Check a11y of linked table rows; fix the accessible label for table row links.
- [x] Elements must have sufficient color contrast (76 violations)
  - Target: `body > .app-container > .sidebar > .sidebar__list > li:nth-of-type(1) > .sidebar__link.sidebar__link--active` (repeat for each blue link)
  - This is the [blue link color we use](https://github.com/thoughtbot/administrate/blob/master/app/assets/stylesheets/administrate/library/_variables.scss#L24)
- [x] Swap `|` in page titles, because screen readers read this as “vertical pipe.”

(I’m using the [aXe Chrome extension](https://chrome.google.com/webstore/detail/axe/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US) to find these errors)

**Comment by pablobm** [comment]
Actually, I'm interested on this one. I think this should be fixed. It's not going to happen today or tomorrow, but we should see to it. In the past I've thought a bit about how this table could be improved for accessibility, and I didn't know this ticket was here. The mention to the aXe tool was useful and I have added it to my browser.

I'm going to reopen this if it's ok. And hopefully we'll look into it. One day...

**Comment by tysongach** [comment]
@pablobm Sounds lovely - I agree we should care about and focus on improving accessibility. I closed because this issue has been open with no movement for so long and I do not have the bandwidth to move it forward I’m also guessing that my list above is now out-of-date (and was non-exhaustive).

**Comment by pangolingo** [comment]
Adding a few to this

- [ ] Sidebar link colors have insufficient contrast against the dark gray background
  - The blue links are ok when on the light gray background of the main area. But the "back to app" link in the sidebar is a little bit lacking on contrast against the darker gray
  - [WebAIM's Contrast Checker](https://webaim.org/resources/contrastchecker/) says it's a contrast ratio of 4.28:1 which is just shy of the recommended 4.5:1
  - We could either darken the blue across the entire app, or just dark blue links in the sidebar
- [ ] Table header cells should not be empty
  - This is more of a best practice: if the table header is empty for a column, it makes it harder for a screenreader user to know what that column is for without navigating through each row
  - We should probably add the text "actions" or something similar to the `<th>` cells for "edit", "delete", etc
- [ ] Voiceover is unable to read table rows
  - In the [demo](https://administrate-demo.herokuapp.com/admin/), Voiceover reads the table as "table, 5 columns, 1 row". It only sees and reads the header row. Other rows are completely inaccessible to this screenreader
  - This is because all the body rows have `role="link"` which removes the "row" role.
  - This is pretty severe!
- [ ] Links must have discernible text
  - In index tables, the contents of each cell is wrapped in an `<a>` tag so it links to the record "show" screen. But for empty columns the `<a>` remains but it's empty. This is confusing to screenreader user
  - This also occurs in cells that have a link already (like the "email" column of the [customer table](https://administrate-demo.herokuapp.com/admin/customers) in the demo

**Comment by pablobm** [comment]
@pangolingo - Would you be up to contribute some fixes? Perhaps the one about the row roles would be a good start?

**Comment by pangolingo** [comment]
> [@pangolingo](https://github.com/pangolingo) - Would you be up to contribute some fixes? Perhaps the one about the row roles would be a good start?

I've started this work here:
- https://github.com/thoughtbot/administrate/pull/2844
- https://github.com/thoughtbot/administrate/pull/2843

**Comment by pangolingo** [comment]
> - Links must have discernible text
>   - In index tables, the contents of each cell is wrapped in an <a> tag so it links to the record "show" screen. But for empty columns the <a> remains but it's empty. This is confusing to screenreader user
>   - This also occurs in cells that have a link already (like the "email" column of the [customer table](https://administrate-demo.herokuapp.com/admin/customers) in the demo

I've looked into this problem (also mentioned by @tysongach in the original report).

The issue is that the collections table wraps each cell in a link. [Reference here](https://github.com/thoughtbot/administrate/blob/main/app/views/administrate/application/_collection.html.erb#L77). But inside that link, each field may be rendered with a link. In particular, emails are wrapped in mailto links, URLs are wrapped in links, and some relationships like belongs_to are wrapped in links. This results in an `<a>` inside an `<a>` which is invalid HTML. Chrome addresses this by splitting it into two side-by-side `<a>` tags, and one of them is empty.

The simplest fix would be to stop wrapping every cell in an `<a>` tag and instead provide a new button to visit the Show page. This would have major ramifications on the UX, so I discarded this idea. 

When rendering the field partial, we could avoid wrapping it in a link if we knew that a partial contained a link. But I doubt that we want to do any kind of pre-parsing of the partial file to find out - that would be cumbersome and ugly. Perhaps instead we could give each field a way to report that it will render a link, so that the collection table can skip wrapping it in a link. One option would be to alter `Administrate::Field::Base` with a property like `should_wrap_index_field_in_link?` that defaults to true. The `Administrate::Field::Email` class could override it to false.

**Comment by pablobm** [comment]
I have had this thought before. My opinion is that we should add an "View" action alongside the existing ones, and use that instead of allowing a whole-row click. Perhaps it's still possible to have a the whole-row click, but as a progressive enhancement, on top of the normal link.

Whether this (assuming you agree with it) should be before or after v1.0, I'll leave to @nickcharlton

**Comment by nickcharlton** [comment]
Hmm, yeah, that's a nice idea. It should be a post-v1.0, but it'd be good to get started on it.
